lib/strings.nix

```
@@ -324,6 +325,65 @@ rec {
   */
   escapeShellArgs = concatMapStringsSep " " escapeShellArg;
 
+  /* Test whether the given name is a valid POSIX shell variable name.
+
+     Type: string -> bool
+
+     Example:
+       isValidShellVar "foo_bar000"
+       => true
+       isValidShellVar "0-bad.jpg"
+       => false
+  */
+  isValidShellVar = name: match "[a-zA-Z_][a-zA-Z0-9_]*" name != null;
+
+  /* Translate a Nix value into a shell variable declaration, with proper escaping.
+
+     Supported value types are strings (mapped to regular variables), lists of strings
+     (mapped to Bash-style arrays) and attribute sets of strings (mapped to Bash-style
+     associative arrays). Note that "strings" include string-coercible values like paths.
+
+     Strings are translated into POSIX sh-compatible code; lists and attribute sets
+     assume a shell that understands Bash syntax (e.g. Bash or ZSH).
+
+     Type: string -> (string | listOf string | attrsOf string) -> string
+
+     Example:
+       ''
+         ${toShellVar foo "some string"}
+         [[ "$foo" == "some string" ]]
+       ''
+  */
+  toShellVar = name: value:
```

Spotted an inconsistency in the usage of "var" only now.

```suggestion
  toShellAssignment = name: value:
```

and similarly elsewhere.

---

I'd like to keep these names short, even if they're not 100% accurate, so that reaching for them isn't a chore.

How about `isValidShellVar` → `isValidShellVariableName` instead to fix the inconsistency?

---

Or `isValidPosixName`, since the reference is https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_235

---

> Or `isValidPosixName`

That's a convenient coincidence.
Let's go with that. It solves the little collision in our terminology here, without long names. I agree "assignment" is a bit cumbersome.